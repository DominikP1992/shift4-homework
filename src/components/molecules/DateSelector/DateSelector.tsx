import IconButton from '@/components/atoms/IconButton/IconButton';
import Text from '@/components/atoms/Text';
import ChevronLeft from '@/components/icons/ChevronLeft';
import ChevronRight from '@/components/icons/ChevronRight';
import { useLocale } from '@/providers/LocaleProvider';
import theme from '@/styles/theme';
import getFormattedMonth from '@/utils/getFormattedMonth';
import isAfterCurrentMonth from '@/utils/isAfterCurrentMonth';
import { x } from '@xstyled/emotion';
import { useEffect, useState } from 'react';

export interface DateSelectorSelectorProps {
  onChange?: (date: Date) => void;
}

export default function DateSelector({ onChange }: DateSelectorSelectorProps) {
  const locale = useLocale();
  const [selectedDate, setSelectedDate] = useState(new Date());

  const handleSelectPrevMonth = () => {
    setSelectedDate((prevDate) => {
      const newDate = new Date(prevDate);
      newDate.setMonth(prevDate.getMonth() - 1);
      return newDate;
    });
  };

  const handleSelectNextMonth = () => {
    setSelectedDate((prevDate) => {
      const newDate = new Date(prevDate);
      newDate.setMonth(prevDate.getMonth() + 1);
      return newDate;
    });
  };

  useEffect(() => {
    onChange && onChange(selectedDate);
  }, [onChange, selectedDate]);

  return (
    <x.div
      borderRadius="4px"
      padding="0px 12px"
      h={theme.sizes.xl}
      backgroundColor="transparent"
      color={theme.colors['purple-gray'] as any}
      border={`1px solid ${theme.colors['input-border']}`}
      display="flex"
      alignItems="center"
      justifyContent="space-between"
      gap="8px"
    >
      <IconButton
        onClick={handleSelectPrevMonth}
        disabled={!isAfterCurrentMonth(selectedDate)}
      >
        <ChevronLeft />
      </IconButton>
      <x.div display="flex" flexDirection="column" alignItems="center">
        <Text as="span" font="Rubik" weight="semibold">
          {getFormattedMonth({ date: selectedDate, locale })}
        </Text>
        <Text as="span" size="sm">
          {selectedDate.getFullYear()}
        </Text>
      </x.div>
      <IconButton onClick={handleSelectNextMonth}>
        <ChevronRight />
      </IconButton>
    </x.div>
  );
}
