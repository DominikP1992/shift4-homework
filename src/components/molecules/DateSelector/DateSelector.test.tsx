import React from 'react';
import '@testing-library/jest-dom';
import DateSelector from '.';
import { fireEvent, render, screen } from '@/tests/testsUtils';

describe('DateSelector', () => {
  const mockDate = new Date('2024-06-15');

  beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(mockDate);
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('renders correctly', () => {
    const { container } = render(<DateSelector onChange={() => {}} />);
    expect(container.firstChild).toBeInTheDocument();
  });

  it('match snapshot', () => {
    const { container } = render(<DateSelector onChange={() => {}} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('allows navigating to the next month', () => {
    render(<DateSelector onChange={() => {}} />);

    const nextMonthButton = screen.getAllByRole('button')[1];

    fireEvent.click(nextMonthButton);
    const nextMonthText = screen.getByText(/July/);
    expect(nextMonthText).toBeInTheDocument();
  });

  it('allows navigating to the previous month', () => {
    render(<DateSelector onChange={() => {}} />);

    const prevMonthButton = screen.getAllByRole('button')[0];
    const nextMonthButton = screen.getAllByRole('button')[1];

    fireEvent.click(nextMonthButton);
    fireEvent.click(nextMonthButton);
    const nextMonthText = screen.getByText(/August/);

    expect(nextMonthText).toBeInTheDocument();

    fireEvent.click(prevMonthButton);
    const prevMonthText = screen.getByText(/July/);
    expect(prevMonthText).toBeInTheDocument();
  });

  it('allows navigating to the previous month', () => {
    render(<DateSelector onChange={() => {}} />);

    const prevMonthButton = screen.getAllByRole('button')[0];
    const nextMonthButton = screen.getAllByRole('button')[1];

    fireEvent.click(nextMonthButton);
    fireEvent.click(nextMonthButton);
    const nextMonthText = screen.getByText(/August/);

    expect(nextMonthText).toBeInTheDocument();

    fireEvent.click(prevMonthButton);
    const prevMonthText = screen.getByText(/July/);
    expect(prevMonthText).toBeInTheDocument();
  });

  it('onChange method returns updated date', () => {
    const mockFn = jest.fn();
    render(<DateSelector onChange={mockFn} />);

    const nextMonthButton = screen.getAllByRole('button')[1];
    fireEvent.click(nextMonthButton);

    const expectedDate = new Date(mockDate);
    expectedDate.setMonth(mockDate.getMonth() + 1);

    expect(mockFn).toHaveBeenCalledWith(expectedDate);
  });
});
