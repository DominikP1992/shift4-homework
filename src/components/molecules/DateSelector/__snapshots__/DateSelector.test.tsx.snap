// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateSelector match snapshot 1`] = `
<div
  class="css-zo0th1"
>
  <button
    aria-label="select prev month"
    class="css-29n6zw"
    disabled=""
  >
    <svg
      fill="none"
      height="12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 12 0 6l6-6 1.4 1.4L2.8 6l4.6 4.6L6 12Z"
        fill="#4D6475"
      />
    </svg>
  </button>
  <div
    class="css-1gpkp3a"
  >
    <span
      class="css-ia5oih"
      weight="semibold"
    >
      June
    </span>
    <span
      class="css-1s199kr"
    >
      2024
    </span>
  </div>
  <button
    aria-label="select next month"
    class="css-29n6zw"
  >
    <svg
      fill="none"
      height="12"
      width="8"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M1.4 12 0 10.6 4.6 6 0 1.4 1.4 0l6 6-6 6Z"
        fill="#4D6475"
      />
    </svg>
  </button>
</div>
`;
